[rfc 2229] Drop fully captured upvars in the same order as the regular drop code

wesleywiser opened the pull request:
Currently, with the new 2021 edition, if a closure captures all of the
fields of an upvar, we'll drop those fields in the order they are used
within the closure instead of the normal drop order (the definition
order of the fields in the type).

This changes that so we sort the captured fields by the definition order
which causes them to drop in that same order as well.

Fixes rust-lang/project-rfc-2229#42

r? @nikomatsakis

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -602,7 +602,47 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            return i1.cmp(&i2);
+                        }
+
+                        (l, r) => bug!("ProjectionKinds were different: ({:?}, {:?})", l, r),
```
a) Might be a good idea to mention why this is a bug
b) I'd prefer if we matched on `(Projection::Index | Subslice, _) | (_,  Projection::Index | Subslice) => bug!("We don't capture index/subslice)`

wesleywiser replied:
I implemented both but in order to make the `match` exhaustive, it's pretty ugly 🙁

arora-aman replied:
Little bit of duplication but this works

```
(l @ ProjectionKind::Slice | ProjectionKind::Index, r)  => bug!(Didn't expect index/slice , l, r)
(l, r  @ ProjectionKind::Slice | ProjectionKind::Index)  => bug!(Didn't expect index/slice , l, r)
(l, r) => bug!("Projections must be of same kind, found", l, r);
```

wesleywiser replied:
I don't think this really gets us very much benefit since we have the wildcard pattern match at the end, if new variants were to be added to `ProjectionKind`, this code would continue to compile. At the same time, we have more code than just falling back to the wildcard match which would show the `ProjectionKind` anyway. I think Niko basically raised the same point [here](https://github.com/rust-lang/rust/pull/89208#discussion_r715630997).

arora-aman commented on src/test/ui/closures/2229_closure_analysis/preserve_field_drop_order2.rs:
```diff
@@ -0,0 +1,28 @@
+// run-pass
+// check-run-results
```
I think we should add some test cases for paths that are multilevel  eg: a.b.c and a.b.d.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3693446439) [(plain)](https://github.com/rust-lang/rust/commit/ab8aef40b1e3a7151daebe7acea2c8b0117a8c9f/checks/3693446439/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 627 error codes
* highest error code: E0785
tidy error: /checkout/src/test/ui/closures/2229_closure_analysis/preserve_field_drop_order2.rs:26: trailing whitespace
tidy error: /checkout/src/test/ui/closures/2229_closure_analysis/preserve_field_drop_order.rs:97: trailing whitespace
Found 0 error codes with no tests
Done!
* 351 features
some tidy checks failed
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:12
```

</details>

arora-aman commented on src/test/ui/closures/2229_closure_analysis/preserve_field_drop_order.rs:
```diff
@@ -0,0 +1,98 @@
+    let a = (HasDrop, HasDrop);
+    let b = (HasDrop, HasDrop);
+
+    let c = #[rustc_capture_analysis]
```
My concern with this test is that we need to reply on checking the stderr, and I feel like if the fields order does get changed, we won't need to update the annoatations in this file and someone might ignore the stderr changes because the ouput needed is all there.

I think we should go with testing with stdout here because changes to stdout are observable behaviour to the user and might be more cause of concern in case something breaks down the line.

wesleywiser replied:
I think it's useful to check both the internal results via `#[rustc_capture_analysis]` as well as the runtime results which is what these two tests do in combination. However, I totally agree with your point that as it is, it would be pretty easy for someone to change this test without realizing what parts are important and which parts aren't. I believe I can remove the checking for the notes we do not care about which should make it much more clear (this is the convention used in other ui tests as well).

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -602,7 +602,47 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        for (_, captures) in &mut root_var_min_capture_list {
+            captures.sort_by(|capture1, capture2| {
+                for (p1, p2) in capture1.place.projections.iter().zip(&capture2.place.projections) {
+                    match (p1.kind, p2.kind) {
```
Might be a good idea to document why we are only comparing the same type here because it might not be apparent to a future reader.

nikomatsakis commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -602,7 +602,47 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        // See https://github.com/rust-lang/project-rfc-2229/issues/42 and
+        // `src/test/ui/closures/2229_closure_analysis/preserve_field_drop_order.rs`.
+        for (_, captures) in &mut root_var_min_capture_list {
+            captures.sort_by(|capture1, capture2| {
```
We *could* use `sort_by_key(|c| &c.place.projections)`, right? It wouldn't detect the buggy cases, but it'd be a heck of a lot shorter!

wesleywiser replied:
`hir::place::Projection` doesn't impl `Ord` so that won't currently work. The MIR version of `Projection` does impl Ord so that doesn't seem entirely unreasonable to do. Would you rather I do that and use `sort_by_key` or leave the code as-is? 🙂

nikomatsakis replied:
I'm of two minds. Thinking about it more, I thought it seems a bit *fragile* to have the runtime semantics be derived by Ord in this way. At minimum, I would want a comment. =)

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -602,7 +602,78 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        ),
+                        (
+                            l
+                            @
```
Not sure why we need to list all of them out explictly

nikomatsakis replied:
I am a fan of listing things explicitly, because it helps detect when new cases are added.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 389365631d3e6d2f533f896635e7768f26359575 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 389365631d3e6d2f533f896635e7768f26359575 -->
<!-- homu: {"type":"Approved","sha":"389365631d3e6d2f533f896635e7768f26359575","approver":"nikomatsakis"} -->

nbdd0121 commented:
Should this be backported to beta?

wesleywiser commented:
Added some comments to existing tests, clarified which notes are important in the `preserve_field_drop_order` test and added some new test cases in `preserve_field_drop_order2`. 

@bors r=nikomatsakis

bors commented:
:pushpin: Commit dd918048a56577dcd7a396065f645a921c3b8e79 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis dd918048a56577dcd7a396065f645a921c3b8e79 -->
<!-- homu: {"type":"Approved","sha":"dd918048a56577dcd7a396065f645a921c3b8e79","approver":"nikomatsakis"} -->

wesleywiser commented:
> Should this be backported to beta?

Yes, I believe we want this to ship with the 2021 edition.

apiraino commented:
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-09-30.20.2354818/near/255583604)

@rustbot label +beta-accepted
